### ISSUE TITLE
[Question] Array type of ROS2 data message, and ROS2 service message

### ISSUE THREAD

**Issue by kteriusghub** [eProsima/SOSS-DDS#47]
Hello ePromisa members!!

I could find which ROS2(or DDS) data types are supported from the current SOSS-DDS(e.g. float32, int64, ...)
I need to send the array data(from sensors)  through SOSS-DDS, but I couldn't find that the array type is supported from the document.
Is the array type supported or not?  If it is not supported, does eProsima have the plan to support the type later?

And, from the sample yaml file, ROS2 topic type can be transferred between two SOSSs.
Does SOSS-DDS support the ROS2 service type? If it doesn't, how about the plan by eProsima?

**Comment by richiware** [comment]
What branch are you testing? `feature/xtypes-dds` branch has a new internal implementation using dynamic types and it should support arrays with any problem.

**Comment by kteriusghub** [comment]
@richiware  
I tested with the `master` branch. Does `feature/xtypes-dds` branch have a so different source stream  from the `master` branch?  I expected that the `master` branch supports the dynamic data types as the `feature/xtypes-dds` branch.
From the `README.md` of SOSS-DDS, I could find the supported data types - boolean, char8, char16, etc.. The md file mentions that the nested structure is also supported, however, I couldn't find the array type is supported or not. So I asked.
If the current implementation supports array type, can you describe the simple example how to use?
Can these comes be the proper one? ( I don't know if std::array can be supported.)

![image](https://user-images.githubusercontent.com/49575160/78089552-c3165a00-7402-11ea-95d4-31f94e259e5a.png)

**Comment by kteriusghub** [comment]
@richiware 
Codes of `feature/xtypes-dds` show that the array and other types are supported. 
But, the branch doesn't contain the dtparser codes and no description about dtparser from CMakeKLists.txt.
So I can't know the notation how to describe those types to the yaml file. Can you explain them?
Additionally, is it possible to overwrite the codes of the branch `feature/xtypes-dds`  to the codes of the master branch?

**Comment by richiware** [comment]
Branch `feature/xtypes-dds` is the next version of SOSS-DDS using DDS X-Types.
In this development we've increased the number of supported types, and also it contains several big changes.

One of these changes is `dtparser` doesn't continue being used to define the types in the YAML file.
In this new version types are defined using IDL format.
```idl
types:
    idls:
        - >
            struct HelloWorld
            {
                string msg[3];
            };
```

Also you can refer IDL files from the YAML configuration:

```idl
    idls:
        - "#include <my_types.idl>"

    paths:
        - "../common-types/"
```

I hope this helps you.

**Comment by kteriusghub** [comment]
@richiware 
Thanks for your kind answer.
Anyway, when I tried to build the source codes, there were a lot of build failure because of '::xtypes'.
There is no 'xtypes' declaration in soss_v2, soss-dds, and ROS2. 
CMake requests xtypes package to build, but there is no  source codes and declaration of 'xtypes'. 
Should I download the eprosima/xtypes from github to remove the build error?
I downloaded it, however, there is no statement of '#include <xtypes/DynamicData.hpp' in soss-dds.
Does 'feature/xtypes-dds' work with soss_v3? If it is true, I don't have source codes of soss_v3.
Can you give the guide about this situation?

**Comment by kteriusghub** [comment]
@richiware 
I found that there are source codes with xtypes from eprosima/soss_v2.git and eprosima/xtypes. And I successfully compiled the soss-dds. The codes support the array variable well.
Can you show any examples how to describe the vector type and the nested structure type in idl format of soss-dds? I tried to add 'typedef' for the nested structure but not worked.

**Comment by kteriusghub** [comment]
@richiware 
I found that the following link explains how to use vector type and the nested structure.
  https://github.com/eProsima/Fast-RTPS/blob/master/include/fastrtps/types/dds-xtypes_typeobject.idl
Anyway, instead of 'types-from', when I mistyped as 'type-from', the result was the segmentation fault with any messages, so I had difficulty.

Can you explain if this SOSS-DDS can support ROS2 service type message?

**Comment by richiware** [comment]
[SOSS](https://github.com/eProsima/SOSS_v2) repository comes with a SOSS-ROS2 SystemHandle which supports specifically ROS2 service type messages. SOSS-DDS supports DDS communication in general. We are working on the documentation. Maybe the current status of the documentation is useful for you:

[Integration Service (future SOSS)](https://integration-service.docs.eprosima.com/en/latest/)
[SOSS](https://soss.docs.eprosima.com/en/latest/index.html)

**Comment by kteriusghub** [comment]
@richiware 
Can you explain more about 'DDS communication in general'? Does it mean that the ROS2 service type message supporting is dependent on SOSS implementation, not SOSS-DDS?
When I added the customized ROS2 service type package to CMakeList.txt of soss-ros2-test, the mix type was not generated (However, the mix file of ROS2 topic type package was successfully generaed)

**Comment by richiware** [comment]
[SOSS-DDS System Handle](https://github.com/eProsima/SOSS-DDS) is used to communicate with DDS applications. It uses directly the [Fast-RTPS](https://github.com/eProsima/Fast-RTPS) API to create publishers and subscribers, and the user defines his types using IDL format.

Although ROS2 uses DDS middlewares, it uses its own types and topics. [SOSS-ROS2 System Handle](https://github.com/eProsima/soss_v2/tree/master/packages/ros2) is prepared to find types in the ROS2 ecosystem and it uses the ROS2 API directly.

Could you explain with more details about your problem with `mix` files? Thanks

**Comment by kteriusghub** [comment]
@richiware 
Thanks for your support!!

I made the ROS2 package with several msg files and srv files, and the package name is 'sosstst_interface'.
And, I added it to CMakeLists.txt of soss/packages/ros2-test as the following.
  -- find_package(sosstst_interface REQUIRED)
  -- soss_rosidl_mix(PACKAGES sosstst_interface   MIDDLEWARES ros2)
  -- target_link_libraries(soss-ros2-test PRIVATE ${sosstst_interface_msgs_LIBRARIES})

And, I added the services section to the yaml file with types-from: ros2  instead of IDL description.
  -- routes:
  ------ros5_to_dds5: {server: ros2, clients: dds }
  -- services:
  -----hello_service_dds: { type: "sosstst_interface/hello_service", route: ros5_to_dds5 }

And I ran SOSS and  SOSS showed the message:
"soss-ros2 could not find .mix file for service type: 
 -- Make sure that you have generated the soss-ros2 extension for that message type by calling soss_rosidl_mix(PACKAGES <package> MIDDLEWARES ros2) in your build system!
soss-ros2 failed to register the required DynamicType [sosstst_interface/GetSerialNo]"

As I already said to you, when I wrote my customized topic of sosstst_interface/laser_scan only, it worked without the error message. This message showed after I added the services section.
Are there any mis-usages for service description of SOSS?

**Comment by richiware** [comment]
SOSS is not finding the `.mix` files. For these cases `SOSS_PREFIX_PATH` environment variable can be used to help SOSS. `export` this environment variable pointing to the location of the *.mix files.

**Comment by kteriusghub** [comment]
@richiware 
I found the reason of "soss-ros2 could not find .mix file for service type:".
Unlike topic type, when I added the services type, `soss/packages/src/Config.cpp` inserts the empty type to the `m_required_types.services`.

`if (it_mw_remap == config.remap.end() || it_mw_remap->second.reply_type == "")
       m_required_types[mw].services.insert(config.reply_type);

`
After I blocked these lines, the error messages was not shown any more.

Anyway, in case of services type, I should write the type with ':request' or ':response'  like the following to the yaml file.
`services:
     hello_service_dds: { type: "sosstst_interface/hello_service:request", route: ros5_to_dds5 }`

If I write the ROS2 service type as the splitted request and response types, SOSS-DDS registeres them as the individual topics, not the integrated one service.
I close this issue because the ROS2-to-ROS2 service operation needs to be more investigated with  both of SOSS and SOSS-DDS, not only SOSS-DDS.
